wallet-new/src/Cardano/Wallet/Kernel/Submission.hs

```diff
@@ -0,0 +1,431 @@
+
+        updateFn :: [Core.TxId] -> Core.TxIn -> [Core.TxId]
+        updateFn acc (Core.TxInUnknown _ _)   = acc
+        updateFn acc (Core.TxInUtxo txHash _) = txHash : acc
+
+
+-- | Overrides the 'Schedule' with an input transaction and a
+-- custom 'Schedule'. Useful to force dispatching, especially in tests.
+overrideSchedule :: WalletSubmission m
+                 -> Slot
+                 -> ScheduledTx
+                 -> WalletSubmission m
+overrideSchedule ws slot scheduled = ws & over (wsState . wssSchedule) override
+    where
+        override :: Schedule -> Schedule
+        override (Schedule s n) =
+            Schedule (IntMap.insertWith mappend (castSlot slot) [scheduled] s) n
+
+
+-- | A getter to the local pending set stored in this 'WalletSubmission'.
+localPendingSet :: Getter (WalletSubmission m) Pending
+localPendingSet = wsState . wssPendingSet
+
+getCurrentSlot :: Getter (WalletSubmission m) Slot
+getCurrentSlot = wsState . wssCurrentSlot
+
+getSchedule :: Getter (WalletSubmission m) Schedule
+getSchedule = wsState . wssSchedule
+
+
+--
+-- Internal API
+--
+
+-- | Schedule the full list of pending transactions.
+-- The transactions will be scheduled immediately in the next 'Slot'.
+schedulePending :: Pending
+                -> WalletSubmission m
+                -> WalletSubmission m
+schedulePending pending ws =
+    let currentSlot = ws ^. wsState . wssCurrentSlot
+    in ws & over (wsState . wssSchedule) (scheduleNext currentSlot)
+    where
+        scheduleNext :: Slot -> Schedule -> Schedule
+        scheduleNext currentSlot (Schedule s n) =
+            let txs     = map toEntry (pending ^. pendingTransactions . fromDb . to M.toList)
+                toEntry (txId, txAux) = ScheduledTx txId txAux (SubmissionCount 0)
+            in Schedule (IntMap.insertWith mappend (castSlot $ mapSlot succ currentSlot) txs s) n
+
+-- | Purges all the scheduled transactions in the given 'Slot'. This is
+-- automatically called by 'tick', and it's responsibility of @rho@ to
+-- correctly reschedule all the transaction.
+purgeSlot :: Slot -> Schedule -> Schedule
+purgeSlot slot (Schedule s n) = Schedule (IntMap.delete (castSlot slot) s) n
+
+-- Ready-to-use 'ResubmissionFunction's.
+
+-- | A 'RetryPolicy' is simply a function which instruct the 'Schedule' when
+-- to attempt resubmitting the given 'ScheduledTx' item.
+type RetryPolicy = SubmissionCount -> Slot -> Maybe Slot
+
+limited :: MaxRetries -> RetryPolicy -> RetryPolicy
+limited maxRetries retryPolicy = \retries currentSlot ->
+    case getSubmissionCount retries >= maxRetries of
+         True  -> Nothing
+         False -> retryPolicy retries currentSlot
+
+type Exponent   = Double
+type MaxRetries = Int
+
+--
+-- Stock retry policies inspired by the 'retry' package.
+--
+
+-- | Very simple policy which merely retries to resubmit the very next 'Slot',
+-- up until 'MaxRetries' attempts.
+constantRetry :: MaxRetries -> RetryPolicy
```

Comment:
Do you think that it might be good to give this function an argument so that `constantRetry n` retires every `n` slot, forcing callers to think about how often they want to retry? Still a bit worried about this ending up in a code base.

Reply:
Fair point -- what if we move `constantRetry` to the test module and we leave only the exponential backoff available? I also agree `constantRetry` is a bit too powerful and I suppose having an `n` parameter won't stop the programmer to have it set to `1`, defeating the purpose. What do you think?

Reply:
(Alternatively we could have something like a `bunnyHopRetry` policy which would retry every other slot, if that's marginally better, but I doubt 😉 )

Reply:
We can't force callers to be sensible, but we can at least nudge them in the right direction :) I think with an argument they at least need to make a conscious decision to retry every single slot.